flink/core/src/main/java/cz/o2/proxima/flink/core/CommitLogSourceFunction.java

```diff
@@ -0,0 +1,320 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.flink.core;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import cz.o2.proxima.direct.commitlog.CommitLogReader;
+import cz.o2.proxima.direct.commitlog.LogObserver;
+import cz.o2.proxima.direct.commitlog.ObserveHandle;
+import cz.o2.proxima.direct.commitlog.Offset;
+import cz.o2.proxima.direct.core.DirectDataOperator;
+import cz.o2.proxima.repository.AttributeDescriptor;
+import cz.o2.proxima.repository.RepositoryFactory;
+import cz.o2.proxima.storage.Partition;
+import cz.o2.proxima.storage.StreamElement;
+import cz.o2.proxima.storage.commitlog.Position;
+import cz.o2.proxima.time.Watermarks;
+import cz.o2.proxima.util.ExceptionUtils;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.flink.api.common.state.CheckpointListener;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.OperatorStateStore;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.runtime.state.JavaSerializer;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;
+import org.apache.flink.streaming.api.watermark.Watermark;
+
+@Slf4j
+public class CommitLogSourceFunction<T> extends RichParallelSourceFunction<T>
+    implements CheckpointListener, CheckpointedFunction {
+
+  private static final String OFFSETS_STATE_NAME = "offsets";
+
+  private static int assignPartition(Partition partition, int numParallelSubtasks) {
+    return partition.getId() % numParallelSubtasks;
+  }
+
+  public static CommitLogSourceFunction<StreamElement> of(
+      RepositoryFactory repositoryFactory, List<AttributeDescriptor<?>> attributeDescriptors) {
+    return new CommitLogSourceFunction<>(
+        repositoryFactory, attributeDescriptors, new IdentityResultExtractor());
+  }
+
+  @FunctionalInterface
+  public interface ResultExtractor<T> extends Serializable {
+
+    T toResult(StreamElement element);
+  }
+
+  public static class IdentityResultExtractor implements ResultExtractor<StreamElement> {
+
+    @Override
+    public StreamElement toResult(StreamElement element) {
+      return element;
+    }
+  }
+
+  @SuppressWarnings("java:S1948")
+  private static class SourceLogObserver<T> implements LogObserver {
```

Comment:
nit: This looks to me like something not exactly related to just commit-log. Maybe more like another top level class because it will be same for other sources. Same as `ResultExtractor`

Reply:
👍 added a test case for this

Reply:
We'll see if this is reusable for implementation of batch sources. Meanwhile this is only place where it's useful, so I'd leave it here for now.